### PR TITLE
panoramic images fixed

### LINKED ISSUE #1815: Panorama thumbnail is pixelized
**Steps to reproduce:** 

1. Scroll featured pictures until you find a very pixelized picture
2. That's it! :-)

All featured pictures are actually high-quality, so it must be possible to generate high-quality thumbnails from them. It might be due to unusual proportions of the picture, example:

Panorama: https://commons.wikimedia.org/wiki/File:Panorama_vom_Gro%C3%9Fen_Leuchtberg.jpg
Result:
screenshot_20180705-000514

Current master

### PR DESCRIPTION
**Pull request by 4D17Y4**
**Description (required)**

Fixes #1815

What changes did you make and why?

Updated image thumb URL depending on the height received.

**Tests performed (required)**

Tested BetaDebug on Nokia 6.1 with API level 29.

### CODE REVIEW

**Comment by 4D17Y4** [comment]
Instead of making another API call, I Figured that we can update the URL which will be the same as if returned from the query.
This reduces the API call as otherwise, we have to call the API multiple times just to get the different image URL.

I understand that this could break if there is some update in the URL returned from the API, but as we have used a similar dependent procedure to get the author of the image I preferred this one.

If this is not acceptable I'll make the API calls to fetch the thumbURL and update the URL.

Up for review @nicolas-raoul , @neslihanturan .
Thanks.

File: data-client/src/main/java/org/wikipedia/gallery/ImageInfo.java [hunk 1]
```diff
@@ -73,4 +90,25 @@ public int getHeight() {
...
+        if(getThumbHeight() < thresholdHeight){
+            // If thumbWidthRetrieved is same as queried width ( If not tells us that the image has no larger dimensions. )
+            if(getThumbWidth() == queryWidth){
+                // Calculate new Width depending on the aspect ratio.
+                final int finalWidth = (int)(thresholdHeight * getThumbWidth() * 1.0 / getThumbHeight());
+                thumbHeight = thresholdHeight;
+                thumbWidth = finalWidth;
+                final String toReplace = "/" + queryWidth + "px";
+                final int position = thumbUrl.lastIndexOf(toReplace);
+                thumbUrl = (new StringBuilder(thumbUrl)).replace(position,position+toReplace.length(),"/"+thumbWidth+"px").toString();
```
Comment (nicolas-raoul) [inline data-client/src/main/java/org/wikipedia/gallery/ImageInfo.java:108]: space after `,`, space before and after `+`

File: data-client/src/main/java/org/wikipedia/gallery/ImageInfo.java [hunk 2]
```diff
@@ -73,4 +90,25 @@ public int getHeight() {
     @Nullable public ExtMetadata getMetadata() {
         return metadata;
     }
+
+    /**
+     * Updates the ThumbUrl if image dimensions are not sufficient.
```
Comment (nicolas-raoul) [inline data-client/src/main/java/org/wikipedia/gallery/ImageInfo.java:95]: Readers might not know what `not sufficient` means, so could you please say it more explicitly? Do not hesitate to explain the specific case of panoramic images.

File: data-client/src/main/java/org/wikipedia/gallery/ImageInfo.java [hunk 3]
```diff
@@ -73,4 +90,25 @@ public int getHeight() {
...
+                // Calculate new Width depending on the aspect ratio.
+                final int finalWidth = (int)(thresholdHeight * getThumbWidth() * 1.0 / getThumbHeight());
+                thumbHeight = thresholdHeight;
+                thumbWidth = finalWidth;
+                final String toReplace = "/" + queryWidth + "px";
+                final int position = thumbUrl.lastIndexOf(toReplace);
+                thumbUrl = (new StringBuilder(thumbUrl)).replace(position,position+toReplace.length(),"/"+thumbWidth+"px").toString();
+            }
+        }
+
```
Comment (nicolas-raoul) [inline data-client/src/main/java/org/wikipedia/gallery/ImageInfo.java:111]: Is this empty line intended?

File: data-client/src/main/java/org/wikipedia/gallery/ImageInfo.java [hunk 4]
```diff
@@ -50,11 +62,16 @@ public int getHeight() {
         return height;
     }
 
+    public int getThumbWidth() { return thumbWidth; }
```
Comment (nicolas-raoul) [inline data-client/src/main/java/org/wikipedia/gallery/ImageInfo.java:65]: javadoc (also for the next one)

File: data-client/src/main/java/org/wikipedia/gallery/ImageInfo.java [hunk 5]
```diff
@@ -29,6 +29,18 @@
...
+    /**
+     * queryWidth, Width parameter of API query in px.
+     */
+    final private int queryWidth = 640;
+
+
+    /**
+     * thresholdHeight, The minimum height of the image in px.
+     */
+    final private int thresholdHeight = 220;
```
Comment (nicolas-raoul) [inline data-client/src/main/java/org/wikipedia/gallery/ImageInfo.java:42]: If it does not change, how about making it `static` maybe?

**Review by nicolas-raoul — CHANGES_REQUESTED** [review]
Great idea to make one less API call!

File: data-client/src/main/java/org/wikipedia/gallery/ImageInfo.java [hunk 6]
```diff
@@ -39,7 +39,7 @@
     /**
      * thresholdHeight, The minimum height of the image in px.
      */
-    final private int thresholdHeight = 220;
+    final private static int thresholdHeight = 220;
```
Comment (nicolas-raoul) [inline data-client/src/main/java/org/wikipedia/gallery/ImageInfo.java:42]: The convention for constants is uppercase: `THRESHOLD_HEIGHT`
This is worth a read: https://www.oracle.com/java/technologies/javase/codeconventions-namingconventions.html :-)

**Review by nicolas-raoul — CHANGES_REQUESTED** [review]

**Comment by nicolas-raoul** [comment]
Strangely the super-high resolution file [Dunes of Amrum](https://commons.wikimedia.org/wiki/File:Dunes_of_Amrum.jpg) is pixelated (even when setting `THRESHOLD_HEIGHT` to `500`):
adb
Any idea why?

**Comment by 4D17Y4** [comment]
Is the issue common to all images?
Or only this particular entry?

Also, do you find any change with this commit in this item's thumbnail?

I'll look into it.
Thanks for notifying

**Comment by nicolas-raoul** [comment]
I notice the low resolution only for this image (not for all images).
It is much worse on master, though.

Would you mind putting breakpoints and checking what thumbnail API calls are made to get the Dunes of Amrum thumbnail?

**Comment by 4D17Y4** [comment]
I've uploaded this same image from the app under test ( sandbox ) , with the name test_panorama and I can reproduce this. 
All the URL seem perfectly fine, the parameters are accurate and clicking on links end up to the high-quality image which its meant to be, but somehow the image view or drawee view is not getting the quality image.

As this is the case with the uploaded image too, I suspect that this may be the case with every image that has width to hight ratio about 15 or more.

I'm looking into the image views attribute that may cause an exception.
If any other image shows this behavior, do notify me. It'll help me confirm that this is due to the dimensions.

Thanks.

**Comment by nicolas-raoul** [comment]
How is the image returned by the API, is it pixelated already?
Or does the pixelation happen afterwards, when displaying it?

**Comment by 4D17Y4** [comment]
I've logged the image url to check for it and here is the result.

Just before calling the function `updateThumbUrl()`
> 2021-04-24 01:55:38.107 30128-30205/fr.free.nrw.commons.beta I/Aditya: Before change https://upload.beta.wmflabs.org/wikipedia/commons/thumb/3/36/Test_panorama.jpg/640px-Test_panorama.jpg


Just after calling the function `updateThumbUrl()`
> 2021-04-24 01:55:38.108 30128-30205/fr.free.nrw.commons.beta I/Aditya: After change https://upload.beta.wmflabs.org/wikipedia/commons/thumb/3/36/Test_panorama.jpg/15238px-Test_panorama.jpg


Before setting it to the imageView in contributions.
> 2021-04-24 01:55:45.070 30128-30128/fr.free.nrw.commons.beta I/Aditya: ImageSource https://upload.beta.wmflabs.org/wikipedia/commons/thumb/3/36/Test_panorama.jpg/15238px-Test_panorama.jpg

U can see the links are perfectly fine with respect to the quality of images.

**Note : I used the threshold as 1000 in this test.**


It might be fresco doing the quality degrade behind the scene.
I'll test it with other methods for loading the image to the view and inform of any update.

Thanks.

**Comment by nicolas-raoul** [comment]
Interesting!
Please do not hesitate to create a new Fresco bug (https://github.com/facebook/fresco/issues I guess?) if there is none about it already. The best is to create a minimal GitHub project that uses just Fresco, and performs only this operation, showing the bug in the smallest project possible.

**Comment by madhurgupta10** [comment]
@4D17Y4 did you try to update Fresco to the latest version?

**Comment by 4D17Y4** [comment]
I did update to the latest version but it didn't help.
I'll check it again 
Thanks.

**Comment by madhurgupta10** [comment]
Fixed in #4467

**Pull request closed without merging**